torchao/quantization/quantize_/workflows/int4/int4_woq_cpu_tensor.py

```diff
@@ -0,0 +1,197 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD 3-Clause license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+from typing import List
+
+import torch
+
+from torchao.quantization.quant_primitives import (
+    MappingType,
+    _choose_qparams_affine_tinygemm,
+    _quantize_affine_tinygemm,
+)
+from torchao.utils import (
+    TorchAOBaseTensor,
+)
+
+__all__ = [
+    "Int4WoqCpuTensor",
+]
+
+aten = torch.ops.aten
+
+
+class Int4WoqCpuTensor(TorchAOBaseTensor):
+    """
+    int4 weight-only quantization on CPU (groupwise quantization only)
+
+    Tensor Attributes:
+        qdata: preshuffled and packed int4 weight, always viewed as a 2D (N, K/2) tensor, last dimension is packed
+               preshuffling is specific to CPU kernels, see Note below.
+        qscale_and_zero: (K/group_size, N, 2), dtype is the same as the original Tensor dtype
+
+    Non-Tensor Attributes:
+        block_size: the block size for quantization, representing the granularity, for groupwise quantization, will have block_size (1, group_size).
+                    we only support group_size = 32/64/128.
+        shape: shape of the original Tensor
+
+    Note on Details for data layout for CPU kernel:
+
+      We use AVX512, AVX512_VNNI and AMX instructions (torch.compile and max-autotune needed for the latter two) to compute GEMM on CPU.
+      For data locality, we preshuffle the data in plain layout (N, K/2) to (N/block_n, K, block_n/2), where block_n = 64. And when packing
+      the last dimension, data are shuffled by lanes before packing two int4 to one int8:
+      block_n = 64 = 16 * 4, so we have 4 lanes, each lane has 16 int4s = [lane0, lane1, lane2, lane3]. We pack them as [lane0|lane2, lane1|lane3].
+      See https://github.com/pytorch/pytorch/blob/32eee8ed225d9f10fbbcb38c24b8b44c24c0c97c/aten/src/ATen/native/cpu/int4mm_kernel.cpp#L583 for more details.
```

Comment:
also if this is based on hardware at the quantization time, what do we do if users quantize the model in one CPU and want to run the model in another CPU?